shotover/src/server.rs

```diff
@@ -233,50 +265,110 @@ impl<C: CodecBuilder + 'static> TcpCodecListener<C> {
                         }
-                        Ok::<(), anyhow::Error>(())
+                        Ok::<bool, anyhow::Error>(false)
                     },
                     // Hot reload request handling
                     hot_reload_request = self.hot_reload_rx.recv() => {
                         if let Some(request) = hot_reload_request{
                             self.handle_hot_reload_request(request).await;
-                            // Wait forever once the FD has been sent. This prevents the loop from continuing
-                            // and attempting to recreate the listener.
-                            // This is fine, since the TcpCodecListener has no more work to do once it has handed off its listener.
-                            // Unfortunately, simply returning from `run` would not work as that would cause shotover to shutdown since there are no more sources running.
-                            futures::future::pending().await
+                            // After handing off the socket FD, mark it as handed off and clear the listener
+                            // This prevents attempting to recreate the listener on the same address
+                            self.socket_handed_off = true;
+                            self.listener = None;
                         }
-                        Ok::<(), anyhow::Error>(())
-                    }
-                }
+                        Ok::<bool, anyhow::Error>(false)
+                    },
+            }
+        }
+        .instrument(span)
+        .await?;
+            if should_return {
+                return Ok(());
             }
-            .instrument(span)
-            .await?;
         }
     }
 
     pub async fn shutdown(&mut self) {
-        join_all(&mut self.connection_handles).await;
+        // First, signal all connections to shutdown
+        for tc in &self.connection_handles {
+            tc.shutdown();
+        }
+
+        // Then wait for all connections to complete
+        join_all(self.connection_handles.iter_mut().map(|tc| &mut tc.handle)).await;
+    }
+
+    pub async fn gradual_shutdown(&mut self) {
+        info!(
+            "[{}] Gradual shutdown initiated - draining connections",
+            self.source_name
+        );
+
+        loop {
+            // Remove finished connections
+            self.connection_handles.retain(|tc| !tc.is_finished());
+
+            let total_connections = self.connection_handles.len();
+
+            if total_connections == 0 {
+                info!(
+                    "[{}] All connections drained, shutting down listener",
+                    self.source_name
+                );
+                return;
+            }
+
+            // Calculate 10% of connections to drain (at least 1 if there are any connections)
+            let connections_to_drain =
+                std::cmp::max(1, (total_connections as f64 * 0.1).ceil() as usize);
```

Comment:
this 10% will shrink every iteration.
Say we have 100 connections.
The first iteration will close 10 connections.
The next iteration will close 9 connections.
And third will close 8 connections. (probably, the math gets more complex at this point)
And so on.

The most predictable behavior would be for each iteration to remove the same amount.
Lets
* move `let total_connections = self.connection_handles.len();` out of the loop, to fix this.
* And then change the `loop { ..; if total_connections == 0 {` into a `while !self.connection_handles.empty()` style loop.